### DISCUSSION TITLE
Is this library secure against XSS ?

### DISCUSSION THREAD

**Discussion by pedro757** [47ng/nuqs#709] [Q&A] [answered]
Just asking

**Comment by franky47** [comment, marked as answer]
If you mean "does it protect you from doing things like this":

```ts
const [redirect] = useQueryState('redirect')
const router = useRouter()

useEffect(() => {
  router.replace(redirect)
}, [redirect])
```

Then no. A little userland logic is needed to prevent XSS and open-redirect vulnerabilities. This can (and probably should) be done in a parser:

```ts
const redirectParser = createParser({
  parse(query) {
    // Very basic check
    if (query.startsWith('/') === false) {
      return null // Possible open redirect vulnerability
    }
    return query
  },
  serialize(value) { return value }
})

function useRedirect() {
  const [redirect] = useQueryState('redirect', redirectParser.withDefault('/'))
  return redirect
}
```

**Reply by abhion** [reply]
@franky47 why does this 

```tsx
http://localhost:3000/en/fle?page_number=1&q=javascript:{alert("ad")}
const [q, setQ] = useQueryState(
    "q",
    parseAsString
      .withDefault(searchParams.get("q") ?? "") // the alert should be injected here
      .withOptions({ throttleMs: 400 }),
  );
```

 not show an alert while

```tsx
const urlToRedirectToOnLogin = searchParams.get("redirect");
 router.push(
        `${urlToRedirectToOnLogin}` ?? DEFAULT_LANDING_PAGE_ON_LOGIN,
 );
```
shows alert when I do `/auth/login?redirect=javascript:{alert("adf")}`

**Reply by franky47** [reply]
In the first case, the `q` variable is set to the string `'javascript:{alert("ad")}'`, there is no evaluation here.

The evaluation (and the XSS / open redirect) comes in the second case, where you connect that variable (whether coming from nuqs or from raw search params) to the router.

As a side note, you don't need to do `.withDefault(searchParams.get("q") ?? "")`, `.withDefault("")` is enough: the default value is returned when the state would otherwise be `null`, ie when there is no search param for the given key in the URL, or its value is invalid for the given parser (here `parseAsString` accepts anything, but more type-restrictive parsers like number or literals may only accept certain values).

**Comment by jjaybrown** [comment]
Should there not ship a parser for this which addresses common xss?

**Reply by franky47** [reply]
Like `parseAsRedirect`? That would be nice to have it built-in yes, but it would only cover open redirects, would you like to open a PR?

For actual XSS (feeding `dangerouslySetInnerHtml` or `eval` stuff from the URL), there isn't much we can do (don't do that).
